### ISSUE TITLE
minimega: refactor vmInfo and friends to an interface

### ISSUE THREAD

**Issue by djfritz** [sandia-minimega/minimega#179]
Trying to think ahead to container support. I think vmInfo needs to change to an interface. A quick look at vminfo.go and vmlist.go indicates we need at least:

type vmInfo interface {
        Start() error
        Stop() error
        Copy() *vmInfo
        String() string
        Info(masks []string) ([]string, error)
        Name() string
        ID() int
        Launch(name string, ack chan int) error
        Kill() error
}

We probably need more once we dig through other obvious dependencies, like the net code. Thoughts on this before I start refactoring?

@jcrussell 
@floren

**Comment by jcrussell** [comment]
Looks good so far. We talked a bit about things that need to happen during the vm "lifecycle". Will those all be handled during the specific functions (e.g. Kill) or do we want to have a (channel) that different components can subscribe to for updates to the VM state?

**Comment by djfritz** [comment]
I'm not sure. We need to refactor the new 'kvm' struct (the old vmInfo) and see how this will work. I'll try and get a first cut of this done by the beginning of next week and put it up for review.

**Comment by djfritz** [comment]
@jcrussell 

Thinking about some future containers and kvm feature that prompted this ticket. The API changes here may influence the implementation somewhat. I'm wondering what the API should look like. Some things are common to both containers and kvm VMs, so we should group things like 'vm stop', 'vm start', 'vm info' (with info showing the vm type, etc). 

What I'm thinking is perhaps something like:

vm stop/start/info/etc...
kvm launch/config foo/etc
container launch/config foo/etc

Or even nesting them:

vm kvm ...

Can you comment on what you think the API here may look like?

**Comment by jcrussell** [comment]
Thoughts:

 * I think we need a subscribe API. Something like:

```

type VMEvent int

const (
	VMStarted VMEvent = 1 << iota
	VMPaused
	VMResumed
	VMKilled
	VMFlushed
)

type vmInfo interface {
	// ...
	Subscribe(chan VMEvent)
}
```

 * Is container the right name? It's a bit verbose. Perhaps lwvm (lightweight VM). lvm would also be a possibility as 'l' and 'k' are next to each other on the keyboard.
 * I think the API should be nested:

```
vm config <kvm,lvm> [true,false]
vm kvm config ...
vm lvm config ...
vm launch
```

**Comment by djfritz** [comment]
Every VM has to subscribe to everything here, which can lead to issues with scale. Let's stick with the simpler solution.

**Comment by djfritz** [comment]
Where is this at? This is blocking container work until we have a known interface.

**Comment by jcrussell** [comment]
Sorry, got distracted by other (shiny) things. The interface is mostly complete [0] but I was planning on fixing #165 at the same time and that'll be a bit more work.

[0] https://github.com/jcrussell/minimega/blob/vmplusplus/src/minimega/vm.go#L44

**Comment by jcrussell** [comment]
Marking this as closed with merge of #229.
